src/sender_queue/dynamic_honey_badger.rs

```diff
@@ -7,27 +7,40 @@ use rand::Rand;
 use serde::{de::DeserializeOwned, Serialize};
 
 use super::{
-    SenderQueue, SenderQueueableDistAlgorithm, SenderQueueableMessage, SenderQueueableOutput,
+    NewValidators, SenderQueue, SenderQueueableDistAlgorithm, SenderQueueableMessage,
+    SenderQueueableOutput,
 };
 use {Contribution, DaStep, NodeIdT};
 
 use dynamic_honey_badger::{
     Batch, Change, ChangeState, DynamicHoneyBadger, Error as DhbError, Message,
 };
 
-impl<C, N> SenderQueueableOutput<N, Message<N>> for Batch<C, N>
+impl<C, N> SenderQueueableOutput<N, (u64, u64)> for Batch<C, N>
 where
     C: Contribution,
     N: NodeIdT + Rand,
 {
-    fn added_peers(&self) -> Vec<N> {
+    fn new_validators(&self) -> NewValidators<N> {
```

Comment:
I've been trying to satisfy this but I think it's impossible for a managed algorithm to provide all participants because it does not know about observers. The new set of validators in `NewValidators::CompletedKeyGen` does not replace the old set of peers but only its subset of old validators which the managed algorithm has to specify because, in turn, the sender queue is not aware of the distinction between validators and observers.

I'd leave `NewValidators` for the reason above. If you know another way, let's continue trying to optimise this.

Reply:
I'll do it a bit differently. Instead of one set, I'll return two: one containing the current participants and one containing the next participants, with an optimisation for the common case when both are the same.